### DISCUSSION TITLE
Unable to plot a contour plot

### DISCUSSION THREAD

**Discussion by ACBNair** [jfree/jfreechart#370] [Q&A]
Hi ,

I been trying to plot a contour with this tool. I have a dataset which plots correctly in plotly.js. 

Tried to put this data as XYZ. Used the sample code available online. But the sample is generating XYZ based on some caluclations.
Following is the code I got online.
        ```
List<String> values = new ArrayList<String>();
    	NumberAxis xAxis = new NumberAxis("X");
        xAxis.setStandardTickUnits(NumberAxis.createIntegerTickUnits());
        xAxis.setLowerMargin(0.0);
        xAxis.setUpperMargin(0.0);
        NumberAxis yAxis = new NumberAxis("Y");
        yAxis.setStandardTickUnits(NumberAxis.createIntegerTickUnits());
        yAxis.setLowerMargin(0.0);
        yAxis.setUpperMargin(0.0);
        
        XYBlockRenderer renderer = new XYBlockRenderer();
        PaintScale scale = new GrayPaintScale(-2.0, 1.0); // Will  be replaced by my color scale
        renderer.setPaintScale(scale);
	
        XYPlot plot = new XYPlot(dataset, xAxis, yAxis, renderer);
        plot.setBackgroundPaint(Color.lightGray);
        plot.setDomainGridlinesVisible(false);
        plot.setRangeGridlinePaint(Color.white);
        JFreeChart chart = new JFreeChart("XYBlockChartDemo1", plot);
        //JFreeChart chart = ChartFactory.createPolarChart("XYBlockChartDemo1", dataset, true, true, true);
        chart.removeLegend();
        chart.setBackgroundPaint(Color.white);
```
But I am getting a chart with individual points, not a heatmap. 

Following is my data : 

 xData = { 145, 108.8, 72.5, 36.3, 0, -36.3, -72.5, -108.8, -145, 0, 0, 0, 0, 0, 0, 0, 0, 102.5,
				76.9, 51.3, 25.6, -25.6, -51.3, -76.9, -102.5, -102.5, -76.9, -51.3, -25.6, 25.6, 51.3, 76.9, 102.5 };
yData = { 0, 0, 0, 0, 0, 0, 0, 0, 0, -145, -108.8, -72.5, -36.3, 36.3, 72.5, 108.8, 145, -102.5,
				-76.9, -51.3, -25.6, 25.6, 51.3, 76.9, 102.5, -102.5, -76.9, -51.3, -25.6, 25.6, 51.3, 76.9, 102.5 };
 zData = { 317.67, 332.13, 343.25, 347.91, 346.29, 347.88, 342.12, 330.75, 316.95, 315.63, 331.06,
				343.08, 349.27, 349.88, 343.64, 330.51, 315.52, 315.98, 332.15, 339.34, 348.35, 347.41, 342.27, 331.82,
				317.4, 318.09, 331.79, 342.95, 347.36, 348.05, 344.09, 329.78, 316.9 };

**Comment by trashgod** [comment]
I don't know `plotly.js`, and it's not clear how you're constructing the `XYZDataset` from the data. You might look among these related [examples](https://stackoverflow.com/search?tab=votes&q=%5bjfreechart%5d%20DefaultXYZDataset&searchOn=3) for reference.

**Reply by ACBNair** [reply]
Thank you for the response. 
Here I will use DefaultXYZDataSet() class to add each of the X[],Y[],and Z[] to it. Then I will pass the same.
But this data is plot as individual dots in the chart not as a contour.

**Reply by trashgod** [reply]
Ah, I see that your data is sparse; I've only used the renderer with continuous data.

**Reply by ACBNair** [reply]
Does it mean we cannot draw it? 🤔

**Reply by trashgod** [reply]
Not at all. The renderer will display the data given to it. Depending on the domain, you can supply the missing data, adding it to your dataset or deriving it in a custom implementation `XYZDataset`.

**Reply by ACBNair** [reply]
Thanks @trashgod . I was able to do that by deriving custom implementation.

I see that the coordinates are drawn using 1 decimal point only. 2.677 ,2.678,2.679 are shown in single place. I was wondering where to change the code as I can see the value is not trimmed anywhere.
